app/gui2/ydoc-server/languageServerSession.ts

```diff
@@ -64,76 +67,128 @@ export class LanguageServerSession extends ObservableV2<Events> {
     this.indexDoc.doc.on('subdocs', (subdocs: { loaded: Set<Y.Doc> }) => {
       for (const doc of subdocs.loaded) {
         const name = this.model.findModuleByDocId(doc.guid)
-        if (name == null) continue
+        if (!name) continue
         const persistence = this.authoritativeModules.get(name)
-        if (persistence == null) continue
+        if (!persistence) continue
       }
     })
+    this.restartClient()
+  }
 
+  static sessions = new Map<string, LanguageServerSession>()
+  static get(url: string): LanguageServerSession {
+    const session = map.setIfUndefined(
+      LanguageServerSession.sessions,
+      url,
+      () => new LanguageServerSession(url),
+    )
+    session.retain()
+    return session
+  }
+
+  private restartClient() {
+    // `this.client` WILL be undefined on the first call.
+    this.client?.close()
+    this.client = createOpenRPCClient(this.url)
+    // `this.ls` WILL be undefined on the first call.
+    this.ls?.destroy()
+    this.ls = new LanguageServer(this.client)
+    this.connection = undefined
```

Comment:
We could just make `setupClient` function with code from line 96 onwards. Then, the constructor could just initialize `client` and `ls` as here, and `restartClient` would do closing/destroying, reinitializing and call `setupClient`. This way we would avoid exclamation marks without being too repetitive.